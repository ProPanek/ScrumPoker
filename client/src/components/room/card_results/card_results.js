import React, {Component} from "react"
import {inject, observer} from "mobx-react";
import Tooltip from '@material-ui/core/Tooltip';
import styled from "styled-components";
import Card from "@material-ui/core/Card";
import Zoom from '@material-ui/core/Zoom';
import ReactCardFlip from 'react-card-flip';
import {decorate, observable, reaction} from "mobx";
import CardTable from "./card_table_dnd"


const Wrapper = styled(Card)`
  display: flex;
  flex-direction: row;
  margin: 50px auto 0;
  width: 90%;
  height: 89px;
  flex-wrap: wrap;
`;

const CardsWrapperWaiting = styled.div`
  &&{
  position: relative;
  display:flex;
  height:89px;
  width: 100%;
  flex-wrap: wrap;
  }
`;
const CardsWrapperResults = styled(CardsWrapperWaiting)`
  &&{
  bottom: 89px;
  }
`;

const StyledCard = styled(Card)`
  &&{
    width:57px;
    height:89px;
    font-size: 30px;
    line-height: 89px;
    background-color:${props => props.color || "#303F9F"};
    color:white;
  }
`;

const CardMargin = styled.div`
  &&{
    margin-right: 20px;
  }
`

class CardResults extends Component {
  constructor(props) {
    super(props)
    this.isFlipped = []
    reaction(
      () => this.props.store.roomStore.cardResults.length,
      () => {
        if (this.props.store.roomStore.cardResults.length > 0) {
          this.startFlipTimer()
        }
      }
    )
  }

  startFlipTimer = () => {
    const {store: {roomStore: {cardResults}}} = this.props
    cardResults.forEach((data, index) => {
      this.isFlipped[index] = false
      setTimeout(() => {
        this.isFlipped[index] = true
      }, 500 * (index + 1))
    })
  }

  render() {
    const {store: {roomStore: {waiting, cardResults, setSelectedCard, selectedCard}}} = this.props
    return (
      <React.Fragment>
        <Wrapper>
          <CardsWrapperWaiting>
            {waiting.map((result, index) => (
              <CardMargin>
                <StyledCard color={"#303F9F"}>
                </StyledCard>
              </CardMargin>
            ))}
            {selectedCard &&
            <CardMargin><StyledCard color={"#303F9F"}>{selectedCard}</StyledCard></CardMargin>}
          </CardsWrapperWaiting>
          <CardsWrapperResults>
            {cardResults.length > 0 && cardResults.map((result, index) => {
              return (
                <ReactCardFlip flipSpeedFrontToBack={0.5} isFlipped={this.isFlipped[index]}>
                  <CardMargin key="front">
                    <StyledCard color={"#303F9F"}>
                    </StyledCard>
                  </CardMargin>
                  <CardMargin key="back">
                    <Tooltip placement="top" TransitionComponent={Zoom} title={result.userName}>
                      <StyledCard color={result.color}>
                        {result.cardValue}
                      </StyledCard>
                    </Tooltip>
                  </CardMargin>
                </ReactCardFlip>
              )
            })
            }
          </CardsWrapperResults>
          <CardTable selectedCard={selectedCard} handleSelectedCard={(value) => {
            setSelectedCard(value)
          }}>
          </CardTable>
        </Wrapper>
      </React.Fragment>
    )
  }
}

decorate(CardResults, {
  isFlipped: observable
});

export {CardResults}
export default inject("store")(observer(CardResults));

