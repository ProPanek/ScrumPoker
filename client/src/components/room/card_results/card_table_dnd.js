import React, {Component} from 'react'
import {
  DropTarget
} from 'react-dnd'

const Types = {
  ITEM: "cardButton"
}

function collect(connect, monitor) {
  return {
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver()
  }
}

const cardTable = {
  drop(props, monitor) {
    if (!monitor.isOver()) {
      return
    }
    if (monitor.didDrop()) {
      return;
    }

    const item = monitor.getItem();
    props.handleSelectedCard(item.value)
  }
};

class CardTable extends Component {
  render() {
    const {connectDropTarget, isOver} = this.props
    const style = {
      backgroundColor: isOver ? "rgba(0,0,0,0.1)" : "rgba(0,0,0,0.0)",
      display: "block",
      position: "relative",
      width: "100%",
      height: "89px",
      top: "-178px",
    }

    return connectDropTarget(
      <div style={style}>
      </div>,
    )
  }
}

export default DropTarget(Types.ITEM, cardTable, collect)(CardTable)
