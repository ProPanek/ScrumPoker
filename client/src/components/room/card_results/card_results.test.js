import React from "react"
import { CardResults } from "./card_results"
import { shallow } from"enzyme";

describe("CardResults", () => {
  it("renders without errors", () => {
    const store = {
      roomStore: { waiting : [], cardResults : []}
    }

    const component = shallow(<CardResults store={store} />)
    expect(component).toMatchSnapshot()
  })
})
