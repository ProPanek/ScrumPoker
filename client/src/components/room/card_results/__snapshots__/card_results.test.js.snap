// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardResults renders without errors 1`] = `
<Fragment>
  <Styled(WithStyles(Card))>
    <styled.div />
    <Styled(styled.div) />
    <DropTarget(CardTable)
      handleSelectedCard={[Function]}
    />
  </Styled(WithStyles(Card))>
</Fragment>
`;
